### ISSUE TITLE
How to force canu -correct to output all reads?

### ISSUE THREAD

**Issue by anitagh** [marbl/canu#1625]
**Canu version 1.8 on CentOS v7.8 Linux OS.**
 
**I'm using canu to correct my Nanopore amplicon reads before feeding them to a rare-event analysis workflow. I don't want to loose any reads.** 

**The reads have minimum of 800 bp overlap with a reference sequence so reducing the  minoverlaplength parameter doesn't affect the result.**

**Canu Command:**
```
canu -correct UseGrid=false -d r2  -p r2  genomeSize=10k corMinCoverage=0 corOutCoverage=20000 corMhapSensitivity=high  -nanopore-raw fastq.fq > r2.log 2>& 1
```
**My fastq file contains 75556 reads. Below is the distribution of the reads.**
```
--   Read length histogram (one '*' equals 527 reads):
--     1000   1999   4778 *********
--     2000   2999  22625 ******************************************
--     3000   3999  36890 **********************************************************************
--     4000   4999   8207 ***************
--     5000   5999   2488 ****
--     6000   6999    302
--     7000   7999     59
--     8000   8999     73
--     9000   9999     50
--    10000  10999     40
--    11000  11999     23
--    12000  12999     11
--    13000  13999      4
--    14000  14999      2
--    15000  15999      1
--    16000  16999      2
--    17000  17999      1
```
**After running the command, I get 75500 reads (loosing 56 reads).**  
```
-- Finished on Thu Feb 20 11:45:24 2020 (one second) with 25198.322 GB free disk space

--                             original      original
--                            raw reads     raw reads
 
--   category                w/overlaps  w/o/overlaps
--   -------------------- ------------- -------------
--   Number of Reads              75501            55
--   Number of Bases          239455808        182883
--   Coverage                 23945.581        18.288
--   Median                        3141          2986
--   Mean                          3171          3325
--   N50                           3370          3035
--   Minimum                       1200          1327
--   Maximum                      17145         10008


 
--
--                                        --------corrected---------  ----------rescued----------
--                             evidence                     expected                     expected
--   category                     reads            raw     corrected            raw     corrected
--   -------------------- -------------  ------------- -------------  ------------- -------------
--   Number of Reads              75556          75500         75500              0             0
--   Number of Bases          239638691      239438829     184537688              0             0
--   Coverage                 23963.869      23943.883     18453.769          0.000         0.000
--   Median                        3141           3141          2333              0             0
--   Mean                          3171           3171          2444              0             0
--   N50                           3369           3370          3119              0             0
--   Minimum                       1200           1200             7              0             0
--   Maximum                      17145          17145         16009              0             0
--
--                        --------uncorrected--------
--                                           expected
--   category                       raw     corrected
--   -------------------- ------------- -------------
--   Number of Reads                 56            56
--   Number of Bases             199862         16240
--   Coverage                    19.986         1.624
--   Median                        2992             0
--   Mean                          3568           290
--   N50                           3048             0
--   Minimum                       1327             0
--   Maximum                      16979         16240
--
--   Maximum Memory           932825762
-- Finished stage 'cor-filterCorrectionLayouts', reset canuIteration.
```

**What parameter do I need to adjust to force all reads to be included?**

**Many Thanks,**
**Azita**

**Comment by skoren** [comment]
You have the right parameters to include all reads, the only reads not being output are the 56 reads which have no overlaps to them even at high sensitivity. There's no way to force correction to output them since they would remain identical to what they were in the input. You could just take those 56 reads as-is if you wanted to keep using them.

**Comment by anitagh** [comment]
Thanks for the response. I'm a bit puzzled. These reads are all mapped to my 1100 bp ref sequence with min of 800 bp overlap. So, they all should have some overlap. (examining the alignment in IGV).  What is the consequence of dropping the minoverlaplength to 300 or 400?

Also, I tested the same cmd with a second dataset. These reads also have 800 bp overlap with the ref sequence. This time 25% of my reads got dropped.

**Comment by skoren** [comment]
I doubt dropping the overlap would help much, even if they map to the reference they might not have a good enough overlap to the other reads (say if it maps to the reference at 25% error). You'd need to try to look at the overlaps report for one of these failing reads, if you share the data we could take a brief look. However, I'd say it's not worth chasing the last 50 reads for correction.

I do see you have slightly more coverage than your corOutCoverage limit, you could try increasing that from 20k to 50k, that might also be causing the 25% drop in your other sample if it is even higher coverage but I can't say for sure because you didn't post that report.

**Comment by anitagh** [comment]
**So, I checked one of the unmapped reads (Read #4)  in the <>.readsToCorrect.log**

![image](https://user-images.githubusercontent.com/24441131/75049819-cf6cd600-5498-11ea-9578-943800fcf04d.png)

**I retrieved the actual id from the readNames.txt and highlighted it in IGV.**

![image](https://user-images.githubusercontent.com/24441131/75049972-1955bc00-5499-11ea-860e-180907132d3e.png)





![image](https://user-images.githubusercontent.com/24441131/75049739-acdabd00-5498-11ea-84cd-17605a35018a.png)


**So, I expect this read to have overlap with others. Just strange...** 

**And increasing corOutCoverage from 20k to 90k didn't change much. I've noticed slight improvement if I drop the genomesize from 10K to 8K. I guess, I follow you advise and just include all uncorrected reads after the correction.**

Thanks,
Azita

**Comment by skoren** [comment]
It's a bit more complicated than just having overlaps since every read can only correct so many other sequences, if everyone picks another read, even if there are some overlaps this read will be ignored. It does look like it is a bit higher error than all the other reads in that area.

**Comment by anitagh** [comment]
Enclosed is the report for the second set. Thanks so much.
[749_r3.log](https://github.com/marbl/canu/files/4237362/749_r3.log)

**Comment by skoren** [comment]
It looks like all the reads here with no overlaps are also due to having no overlaps, it would be interesting to see why they are being excluded. Are you able to share this second dataset, the FAQ has instructions on how to send us data.

**Comment by anitagh** [comment]
Hi Sergy - I need to check with my company to see if I can upload the sample dataset. But I ran a simple test. Extracted all uncorrected reads from the second dataset and re-submitted to the canu.  81% of those reads got corrected. So, it seems I need to do a couple iterations to get most of the reads corrected.
[749_uncorrect.log](https://github.com/marbl/canu/files/4239636/749_uncorrect.log)

Best,
Azita
